### PR TITLE
Some fixes or not

### PR DESCRIPTION
**Pull request by IRacle1**
e
- again fixed unity objects comparing in `Item::Get` (that was reason cringey crushes on my server)
- optimization and maybe fixes for Shooting and firearm-requests events (plus i changed ctors a lil bit, technically a breaking change)
- some todos
- removed some `ChangingMoveState` event functionality, cuz it will desync player
- fixed server crush in `Scp106::Attacking` patch

### CODE REVIEW

File: Exiled.Events/EventArgs/Player/AimingDownSightEventArgs.cs [hunk 1]
```diff
@@ -23,19 +24,18 @@ public class AimingDownSightEventArgs : IPlayerEvent, IFirearmEvent
...
+        ///     <inheritdoc cref="Firearm" />
+        /// </param>
         /// <param name="adsIn">
         ///     <inheritdoc cref="AdsIn" />
         /// </param>
         /// <param name="adsOut">
         ///     <inheritdoc cref="AdsOut" />
         /// </param>
-        public AimingDownSightEventArgs(Player player, bool adsIn, bool adsOut)
+        public AimingDownSightEventArgs(Player player, Firearm firearm, bool adsIn, bool adsOut)
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Player/AimingDownSightEventArgs.cs:36]: That's a breaking change, we're changing the signature of a `public` constructor.

File: Exiled.Events/EventArgs/Player/DryfiringWeaponEventArgs.cs [hunk 2]
```diff
@@ -23,12 +23,15 @@ public class DryfiringWeaponEventArgs : IPlayerEvent, IFirearmEvent, IDeniableEv
...
         ///     <inheritdoc cref="Player" />
         /// </param>
+        /// <param name="firearm">
+        ///     <inheritdoc cref="Firearm" />
+        /// </param>
         /// <param name="isAllowed">
         ///     <inheritdoc cref="IsAllowed" />
         /// </param>
-        public DryfiringWeaponEventArgs(Player player, bool isAllowed = true)
+        public DryfiringWeaponEventArgs(Player player, Firearm firearm, bool isAllowed = true)
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Player/DryfiringWeaponEventArgs.cs:32]: Same as above.

File: Exiled.Events/EventArgs/Player/ReloadingWeaponEventArgs.cs [hunk 3]
```diff
@@ -23,12 +23,15 @@ public class ReloadingWeaponEventArgs : IPlayerEvent, IFirearmEvent, IDeniableEv
...
         ///     <inheritdoc cref="Player" />
         /// </param>
+        /// <param name="firearm">
+        ///     <inheritdoc cref="Firearm" />
+        /// </param>
         /// <param name="isAllowed">
         ///     <inheritdoc cref="IsAllowed" />
         /// </param>
-        public ReloadingWeaponEventArgs(Player player, bool isAllowed = true)
+        public ReloadingWeaponEventArgs(Player player, Firearm firearm, bool isAllowed = true)
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Player/ReloadingWeaponEventArgs.cs:32]: Same as above.

File: Exiled.Events/EventArgs/Player/ShootingEventArgs.cs [hunk 4]
```diff
@@ -28,12 +30,16 @@ public class ShootingEventArgs : IPlayerEvent, IDeniableEvent
...
         ///     <inheritdoc cref="Player" />
         /// </param>
+        /// <param name="firearm">
+        ///     <inheritdoc cref="Firearm" />
+        /// </param>
         /// <param name="msg">
         ///     <inheritdoc cref="ShotMessage" />
         /// </param>
-        public ShootingEventArgs(Player shooter, ShotMessage msg)
+        public ShootingEventArgs(Player shooter, Firearm firearm, ShotMessage msg)
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Player/ShootingEventArgs.cs:39]: Same as above.

File: Exiled.Events/EventArgs/Player/TogglingWeaponFlashlightEventArgs.cs [hunk 5]
```diff
@@ -23,26 +23,27 @@ public class TogglingWeaponFlashlightEventArgs : IPlayerEvent, IFirearmEvent, ID
...
+        ///     <inheritdoc cref="Firearm" />
+        /// </param>
         /// <param name="newState">
         ///     <inheritdoc cref="NewState" />
         /// </param>
         /// <param name="isAllowed">
         ///     <inheritdoc cref="IsAllowed" />
         /// </param>
-        public TogglingWeaponFlashlightEventArgs(Player player, bool newState, bool isAllowed = true)
+        public TogglingWeaponFlashlightEventArgs(Player player, Firearm firearm, bool newState, bool isAllowed = true)
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Player/TogglingWeaponFlashlightEventArgs.cs:35]: Same as above.

File: Exiled.Events/EventArgs/Player/UnloadingWeaponEventArgs.cs [hunk 6]
```diff
@@ -23,12 +23,15 @@ public class UnloadingWeaponEventArgs : IPlayerEvent, IFirearmEvent, IDeniableEv
...
         ///     <inheritdoc cref="Player" />
         /// </param>
+        /// <param name="firearm">
+        ///     <inheritdoc cref="Firearm" />
+        /// </param>
         /// <param name="isAllowed">
         ///     <inheritdoc cref="IsAllowed" />
         /// </param>
-        public UnloadingWeaponEventArgs(Player player, bool isAllowed = true)
+        public UnloadingWeaponEventArgs(Player player, Firearm firearm, bool isAllowed = true)
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Player/UnloadingWeaponEventArgs.cs:32]: Same as above.

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]

File: Exiled.Events/EventArgs/Player/ChangingMoveStateEventArgs.cs [hunk 7]
```diff
@@ -54,11 +54,13 @@ public ChangingMoveStateEventArgs(Player player, PlayerMovementState oldState, P
         /// <summary>
         ///     Gets or sets the new state.
         /// </summary>
+        // TODO: remove setter
```
Comment (louis1706) [inline Exiled.Events/EventArgs/Player/ChangingMoveStateEventArgs.cs:57]: use Obsolete("Will desync client")

File: Exiled.Events/EventArgs/Player/ChangingMoveStateEventArgs.cs [hunk 8]
```diff
@@ -54,11 +54,13 @@ public ChangingMoveStateEventArgs(Player player, PlayerMovementState oldState, P
...
         public PlayerMovementState NewState { get; set; }
 
         /// <summary>
         ///     Gets or sets a value indicating whether the player can change the movement state.
         /// </summary>
+        // TODO: remove
```
Comment (louis1706) [inline Exiled.Events/EventArgs/Player/ChangingMoveStateEventArgs.cs:63]: use Obsolete("Will desync client")

File: Exiled.Events/EventArgs/Player/ChangingMoveStateEventArgs.cs [hunk 9]
```diff
@@ -54,11 +57,19 @@ public ChangingMoveStateEventArgs(Player player, PlayerMovementState oldState, P
         /// <summary>
         ///     Gets or sets the new state.
         /// </summary>
-        public PlayerMovementState NewState { get; set; }
+        // TODO: remove setter
+        public PlayerMovementState NewState
+        {
+            get;
+            [Obsolete("Functional was removed due desync problems")]
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Player/ChangingMoveStateEventArgs.cs:64]: `Setters was removed due to desync problems.`

File: Exiled.Events/EventArgs/Player/ChangingMoveStateEventArgs.cs [hunk 10]
```diff
@@ -54,11 +57,19 @@ public ChangingMoveStateEventArgs(Player player, PlayerMovementState oldState, P
...
+            set;
+        }
 
         /// <summary>
         ///     Gets or sets a value indicating whether the player can change the movement state.
         /// </summary>
-        public bool IsAllowed { get; set; }
+        // TODO: remove
+        [Obsolete("Functional was removed due desync problems")]
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Player/ChangingMoveStateEventArgs.cs:72]: Same as above.

File: Exiled.Events/EventArgs/Player/TogglingWeaponFlashlightEventArgs.cs [hunk 11]
```diff
@@ -23,26 +23,27 @@ public class TogglingWeaponFlashlightEventArgs : IPlayerEvent, IFirearmEvent, ID
...
+            Firearm = firearm;
             Player = player;
             NewState = newState;
             IsAllowed = isAllowed;
         }
 
-#pragma warning disable SA1623 // Property summary documentation should match accessors
         /// <summary>
-        ///     Gets or sets the new weapon's flashlight state.
+        ///     Gets or sets a value indicating whether or not the new weapon's flashlight state will be enabled or not.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Player/TogglingWeaponFlashlightEventArgs.cs:44]: `Gets or sets a value indicating whether the new weapon's flashlight state will be enabled.`

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]

File: Exiled.Events/EventArgs/Player/TogglingWeaponFlashlightEventArgs.cs (hunk 11 above)
Comment (IRacle1) [reply Exiled.Events/EventArgs/Player/TogglingWeaponFlashlightEventArgs.cs]: fixed

File: Exiled.Events/EventArgs/Player/ChangingMoveStateEventArgs.cs (hunk 10 above)
Comment (IRacle1) [reply Exiled.Events/EventArgs/Player/ChangingMoveStateEventArgs.cs]: fixed

File: Exiled.Events/EventArgs/Player/ChangingMoveStateEventArgs.cs (hunk 9 above)
Comment (IRacle1) [reply Exiled.Events/EventArgs/Player/ChangingMoveStateEventArgs.cs]: fixed

**Pull request merged**